// ng-apexcharts/esm5/lib/ng-apexcharts.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ng-apexcharts.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { NgModule } from "@angular/core";
import { ChartComponent } from "./chart/chart.component";
import ApexCharts from "apexcharts/dist/apexcharts.esm.js";
import * as ɵngcc0 from '@angular/core';
window.ApexCharts = ApexCharts;
/** @type {?} */
var declerations = [ChartComponent];
var NgApexchartsModule = /** @class */ (function () {
    function NgApexchartsModule() {
    }
NgApexchartsModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgApexchartsModule });
NgApexchartsModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgApexchartsModule_Factory(t) { return new (t || NgApexchartsModule)(); }, imports: [[]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgApexchartsModule, { declarations: [ChartComponent], exports: [ChartComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgApexchartsModule, [{
        type: NgModule,
        args: [{
                declarations: tslib_1.__spread(declerations),
                imports: [],
                exports: tslib_1.__spread(declerations)
            }]
    }], function () { return []; }, null); })();
    return NgApexchartsModule;
}());
export { NgApexchartsModule };
