// ngx-drag-drop/esm5/dnd.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { DndDraggableDirective, DndDragImageRefDirective } from "./dnd-draggable.directive";
import { DndDropzoneDirective, DndPlaceholderRefDirective } from "./dnd-dropzone.directive";
import { DndHandleDirective } from "./dnd-handle.directive";
import * as ɵngcc0 from '@angular/core';
var DndModule = /** @class */ (function () {
    function DndModule() {
    }
DndModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: DndModule });
DndModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function DndModule_Factory(t) { return new (t || DndModule)(); }, imports: [[
            CommonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DndModule, { declarations: function () { return [DndDraggableDirective,
        DndDropzoneDirective,
        DndHandleDirective,
        DndPlaceholderRefDirective,
        DndDragImageRefDirective]; }, imports: function () { return [CommonModule]; }, exports: function () { return [DndDraggableDirective,
        DndDropzoneDirective,
        DndHandleDirective,
        DndPlaceholderRefDirective,
        DndDragImageRefDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DndModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    DndDraggableDirective,
                    DndDropzoneDirective,
                    DndHandleDirective,
                    DndPlaceholderRefDirective,
                    DndDragImageRefDirective
                ],
                exports: [
                    DndDraggableDirective,
                    DndDropzoneDirective,
                    DndHandleDirective,
                    DndPlaceholderRefDirective,
                    DndDragImageRefDirective
                ]
            }]
    }], function () { return []; }, null); })();
    return DndModule;
}());
export { DndModule };
