// angular-archwizard/esm5/lib/util/wizard-step.interface.js
import { __decorate, __metadata } from "tslib";
import { ContentChild, EventEmitter, HostBinding, Input, Output, Directive } from '@angular/core';
import { WizardStepSymbolDirective } from '../directives/wizard-step-symbol.directive';
import { WizardStepTitleDirective } from '../directives/wizard-step-title.directive';
/**
 * Basic functionality every type of wizard step needs to provide
 *
 * @author Marc Arndt
 */
import * as ɵngcc0 from '@angular/core';
var WizardStep = /** @class */ (function () {
    function WizardStep() {
        /**
         * A symbol property, which contains an optional symbol for the step inside the navigation bar.
         * Takes effect when `stepSymbolTemplate` is not defined or null.
         */
        this.navigationSymbol = { symbol: '' };
        /**
         * A boolean describing if the wizard step is currently selected
         */
        this.selected = false;
        /**
         * A boolean describing if the wizard step has been completed
         */
        this.completed = false;
        /**
         * A boolean describing if the wizard step is shown as completed when the wizard is presented to the user
         *
         * Users will typically use `CompletedStepDirective` to set this flag
         */
        this.initiallyCompleted = false;
        /**
         * A boolean describing if the wizard step is being edited after being competed
         *
         * This flag can only be true when `selected` is true.
         */
        this.editing = false;
        /**
         * A boolean describing, if the wizard step should be selected by default, i.e. after the wizard has been initialized as the initial step
         */
        this.defaultSelected = false;
        /**
         * A boolean describing if the wizard step is an optional step
         */
        this.optional = false;
        /**
         * A function or boolean deciding, if this step can be entered
         */
        this.canEnter = true;
        /**
         * A function or boolean deciding, if this step can be exited
         */
        this.canExit = true;
        /**
         * This [[EventEmitter]] is called when the step is entered.
         * The bound method should be used to do initialization work.
         */
        this.stepEnter = new EventEmitter();
        /**
         * This [[EventEmitter]] is called when the step is exited.
         * The bound method can be used to do cleanup work.
         */
        this.stepExit = new EventEmitter();
    }
    WizardStep_1 = WizardStep;
    Object.defineProperty(WizardStep.prototype, "hidden", {
        /**
         * Returns true if this wizard step should be visible to the user.
         * If the step should be visible to the user false is returned, otherwise true
         */
        get: function () {
            return !this.selected;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * This method returns true, if this wizard step can be transitioned with a given direction.
     * Transitioned in this case means either entered or exited, depending on the given `condition` parameter.
     *
     * @param condition A condition variable, deciding if the step can be transitioned
     * @param direction The direction in which this step should be transitioned
     * @returns A [[Promise]] containing `true`, if this step can transitioned in the given direction
     * @throws An `Error` is thrown if `condition` is neither a function nor a boolean
     */
    WizardStep.canTransitionStep = function (condition, direction) {
        if (typeof (condition) === typeof (true)) {
            return Promise.resolve(condition);
        }
        else if (condition instanceof Function) {
            return Promise.resolve(condition(direction));
        }
        else {
            return Promise.reject(new Error("Input value '" + condition + "' is neither a boolean nor a function"));
        }
    };
    /**
     * A function called when the step is entered
     *
     * @param direction The direction in which the step is entered
     */
    WizardStep.prototype.enter = function (direction) {
        this.stepEnter.emit(direction);
    };
    /**
     * A function called when the step is exited
     *
     * @param direction The direction in which the step is exited
     */
    WizardStep.prototype.exit = function (direction) {
        this.stepExit.emit(direction);
    };
    /**
     * This method returns true, if this wizard step can be entered from the given direction.
     * Because this method depends on the value `canEnter`, it will throw an error, if `canEnter` is neither a boolean
     * nor a function.
     *
     * @param direction The direction in which this step should be entered
     * @returns A [[Promise]] containing `true`, if the step can be entered in the given direction, false otherwise
     * @throws An `Error` is thrown if `anEnter` is neither a function nor a boolean
     */
    WizardStep.prototype.canEnterStep = function (direction) {
        return WizardStep_1.canTransitionStep(this.canEnter, direction);
    };
    /**
     * This method returns true, if this wizard step can be exited into given direction.
     * Because this method depends on the value `canExit`, it will throw an error, if `canExit` is neither a boolean
     * nor a function.
     *
     * @param direction The direction in which this step should be left
     * @returns A [[Promise]] containing `true`, if the step can be exited in the given direction, false otherwise
     * @throws An `Error` is thrown if `canExit` is neither a function nor a boolean
     */
    WizardStep.prototype.canExitStep = function (direction) {
        return WizardStep_1.canTransitionStep(this.canExit, direction);
    };
    var WizardStep_1;
    __decorate([
        ContentChild(WizardStepTitleDirective),
        __metadata("design:type", WizardStepTitleDirective)
    ], WizardStep.prototype, "stepTitleTemplate", void 0);
    __decorate([
        ContentChild(WizardStepSymbolDirective),
        __metadata("design:type", WizardStepSymbolDirective)
    ], WizardStep.prototype, "stepSymbolTemplate", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], WizardStep.prototype, "stepId", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], WizardStep.prototype, "stepTitle", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], WizardStep.prototype, "navigationSymbol", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], WizardStep.prototype, "canEnter", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], WizardStep.prototype, "canExit", void 0);
    __decorate([
        Output(),
        __metadata("design:type", EventEmitter)
    ], WizardStep.prototype, "stepEnter", void 0);
    __decorate([
        Output(),
        __metadata("design:type", EventEmitter)
    ], WizardStep.prototype, "stepExit", void 0);
    __decorate([
        HostBinding('hidden'),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [])
    ], WizardStep.prototype, "hidden", null);
WizardStep.ɵfac = function WizardStep_Factory(t) { return new (t || WizardStep)(); };
WizardStep.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: WizardStep, contentQueries: function WizardStep_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, WizardStepTitleDirective, true);
        ɵngcc0.ɵɵcontentQuery(dirIndex, WizardStepSymbolDirective, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.stepTitleTemplate = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.stepSymbolTemplate = _t.first);
    } }, hostVars: 1, hostBindings: function WizardStep_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵhostProperty("hidden", ctx.hidden);
    } }, inputs: { navigationSymbol: "navigationSymbol", canEnter: "canEnter", canExit: "canExit", stepId: "stepId", stepTitle: "stepTitle" }, outputs: { stepEnter: "stepEnter", stepExit: "stepExit" } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(WizardStep, [{
        type: Directive
    }], function () { return []; }, { navigationSymbol: [{
            type: Input
        }], canEnter: [{
            type: Input
        }], canExit: [{
            type: Input
        }], stepEnter: [{
            type: Output
        }], stepExit: [{
            type: Output
        }], hidden: [{
            type: HostBinding,
            args: ['hidden']
        }], stepTitleTemplate: [{
            type: ContentChild,
            args: [WizardStepTitleDirective]
        }], stepSymbolTemplate: [{
            type: ContentChild,
            args: [WizardStepSymbolDirective]
        }], stepId: [{
            type: Input
        }], stepTitle: [{
            type: Input
        }] }); })();
    return WizardStep;
}());
export { WizardStep };
