// ng-click-outside/lib_commonjs/click-outside.module.js
"use strict";
var ɵngcc0 = require('@angular/core');
var ɵngcc1 = require('./click-outside.directive');
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var click_outside_directive_1 = require("./click-outside.directive");
var ClickOutsideModule = (function () {
    function ClickOutsideModule() {
    }
ClickOutsideModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClickOutsideModule });
ClickOutsideModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClickOutsideModule_Factory(t) { return new (t || ClickOutsideModule)(); } });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClickOutsideModule, { declarations: [ɵngcc1.ClickOutsideDirective], exports: [ɵngcc1.ClickOutsideDirective] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClickOutsideModule, [{
        type: core_1.NgModule,
        args: [{
                declarations: [click_outside_directive_1.ClickOutsideDirective],
                exports: [click_outside_directive_1.ClickOutsideDirective]
            }]
    }], function () { return []; }, null); })();
    return ClickOutsideModule;
}());
exports.ClickOutsideModule = ClickOutsideModule;
